File: runners/go/emit.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package main

import (
	"fmt"
	"strconv"

	"github.com/apache/iceberg-go"
	"gopkg.in/yaml.v3"
)

// CanonicalOutput is the runner's emit shape (expected-output.schema.json). A
// blessed run of this output becomes a golden.
type CanonicalOutput struct {
	SpecID       string        `json:"spec-id"`
	Accept       bool          `json:"accept"`
	Snapshots    []SnapshotOut `json:"snapshots,omitempty"`
	Observations []Observation `json:"observations"`
}

// SnapshotOut is one commit-ordered snapshot with only spec-pinned, layout-
// independent facts. The random snapshot-id is replaced by the commit ordinal.
type SnapshotOut struct {
	Ordinal        int          `json:"ordinal"`
	Parent         *int         `json:"parent"`
	Operation      string       `json:"operation"`
	SequenceNumber *int64       `json:"sequence-number,omitempty"`
	Summary        *SummaryOut  `json:"summary,omitempty"`
	DeleteFiles    []DeleteFile `json:"delete-files,omitempty"`
}

// SummaryOut carries only layout-independent aggregate metrics.
type SummaryOut struct {
	TotalRecords     *int `json:"total-records,omitempty"`
	AddedRecords     *int `json:"added-records,omitempty"`
	DeletedRecords   *int `json:"deleted-records,omitempty"`
	TotalDeleteFiles *int `json:"total-delete-files,omitempty"`
}

// DeleteFile is a spec-pinned fact about a delete file added in a snapshot.
type DeleteFile struct {
	Content int    `json:"content"` // 1=positional (incl. DV), 2=equality
	Format  string `json:"format,omitempty"`
}

// Observation is a decoded read of the table at one observe point.
type Observation struct {
	At            any                     `json:"at"`
	Bind          string                  `json:"bind,omitempty"`
	IcebergSchema []schemaField           `json:"iceberg-schema"`
	DecodedRows   []map[string]*valueNode `json:"decoded-rows"`
}

// summaryOut extracts the layout-independent aggregate metrics from a
// snapshot's summary properties (all string-valued in iceberg-go). Missing
// keys are left nil so they are omitted from output.
func summaryOut(props iceberg.Properties) *SummaryOut {
	s := &SummaryOut{}
	s.TotalRecords = intProp(props, "total-records")
	s.AddedRecords = intProp(props, "added-records")
	s.DeletedRecords = intProp(props, "deleted-records")
	// total-delete-files is only meaningful once deletes exist; the goldens
	// omit it on append snapshots, so suppress a zero value.
	if tdf := intProp(props, "total-delete-files"); tdf != nil && *tdf > 0 {
		s.TotalDeleteFiles = tdf
	}
	if s.TotalRecords == nil && s.AddedRecords == nil && s.DeletedRecords == nil && s.TotalDeleteFiles == nil {
		return nil
	}
	return s
}

func intProp(props iceberg.Properties, key string) *int {
	v, ok := props[key]
	if !ok {
		return nil
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return nil
	}
	return &n
}

// observePoint is the resolved target of an observe entry.
type observeKind int

const (
	atLatest  observeKind = iota
	atBind                // time-travel to a bound observation
	atOrdinal             // time-travel to a commit ordinal
)

type observePoint struct {
	kind    observeKind
	value   any    // the value echoed into output ('latest', a bind name, or an ordinal)
	bind    string // bind name when kind==atBind
	ordinal int    // ordinal when kind==atOrdinal
}

// observeAt resolves an observe entry's 'at' node.
func observeAt(e Entry) (observePoint, error) {
	node := &e.At
	if node.Kind == 0 {
		return observePoint{}, fmt.Errorf("observe missing 'at'")
	}
	if node.Kind == yaml.ScalarNode {
		if node.Tag == "!!int" {
			n, err := strconv.Atoi(node.Value)
			if err != nil {
				return observePoint{}, err
			}
			return observePoint{kind: atOrdinal, value: n, ordinal: n}, nil
		}
		if node.Value == "latest" {
			return observePoint{kind: atLatest, value: "latest"}, nil
		}
		return observePoint{kind: atBind, value: node.Value, bind: node.Value}, nil
	}
	return observePoint{}, fmt.Errorf("unsupported 'at' node")
}

File: runners/rust/src/emit.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! Canonical output shape (expected-output.schema.json). A blessed run of this
//! output becomes a golden. Serialized with serde_json; `#[serde(skip_serializing_if)]`
//! keeps optional facts out when absent so the shape matches the corpus goldens.

use std::collections::BTreeMap;

use serde::Serialize;

/// The runner's top-level emit shape.
#[derive(Debug, Serialize)]
pub struct CanonicalOutput {
    #[serde(rename = "spec-id")]
    pub spec_id: String,
    pub accept: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub snapshots: Vec<SnapshotOut>,
    pub observations: Vec<Observation>,
}

/// One commit-ordered snapshot with only spec-pinned, layout-independent facts.
/// The random snapshot-id is replaced by the commit ordinal.
#[derive(Debug, Serialize)]
pub struct SnapshotOut {
    pub ordinal: i64,
    pub parent: Option<i64>,
    pub operation: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<SummaryOut>,
    #[serde(rename = "delete-files", skip_serializing_if = "Vec::is_empty")]
    pub delete_files: Vec<DeleteFile>,
}

/// Layout-independent aggregate metrics.
#[derive(Debug, Serialize)]
pub struct SummaryOut {
    #[serde(rename = "total-records", skip_serializing_if = "Option::is_none")]
    pub total_records: Option<i64>,
    #[serde(rename = "added-records", skip_serializing_if = "Option::is_none")]
    pub added_records: Option<i64>,
    #[serde(rename = "deleted-records", skip_serializing_if = "Option::is_none")]
    pub deleted_records: Option<i64>,
    #[serde(rename = "total-delete-files", skip_serializing_if = "Option::is_none")]
    pub total_delete_files: Option<i64>,
}

/// A spec-pinned fact about a delete file added in a snapshot.
#[derive(Debug, Serialize)]
pub struct DeleteFile {
    pub content: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<String>,
}

/// A decoded read of the table at one observe point.
#[derive(Debug, Serialize)]
pub struct Observation {
    pub at: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bind: Option<String>,
    #[serde(rename = "iceberg-schema")]
    pub iceberg_schema: Vec<SchemaField>,
    #[serde(rename = "decoded-rows")]
    pub decoded_rows: Vec<DecodedRow>,
}

/// One entry of an observation's iceberg-schema array.
#[derive(Debug, Serialize)]
pub struct SchemaField {
    #[serde(rename = "field-id")]
    pub field_id: i32,
    pub name: String,
    #[serde(rename = "type")]
    pub type_name: String,
}

/// One decoded row: field-id (as string key) -> value node. Row order is not
/// significant (the orchestrator sorts by __rowkey); null cells are omitted.
pub type DecodedRow = BTreeMap<String, ValueNode>;

/// PR #2 type-annotated value tree. Primitive: {type, value}. (Nested variants
/// are a Phase 4 extension.) 64-bit integers are serialized as JSON strings.
#[derive(Debug, Serialize)]
pub struct ValueNode {
    #[serde(rename = "type")]
    pub type_name: String,
    pub value: serde_json::Value,
}
